### PR TITLE
feat: apply prepaid credits

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Apply prepaid credits on subscription invoices

## Description

- Created `applied_prepaid_credit` collection that creates relation between invoice and wallet transaction
- Apply prepaid credit on invoice
- Update slim template

### CODE REVIEW

File: app/services/invoices/create_service.rb [hunk 1]
```diff
@@ -123,12 +124,26 @@ def applied_coupon
...
     def should_create_credit?
       return false if applied_coupon.nil?
 
       applied_coupon.amount_currency == currency
     end
 
+    def should_create_applied_prepaid_credit?
+      return false unless wallet
+      return false unless wallet.active?
+      return false unless BigDecimal(wallet.balance) > 0
```
Comment (vincent-pochet) [inline app/services/invoices/create_service.rb:142]: Rails should handle the data type automatically :)

File: app/services/invoices/create_service.rb [hunk 2]
```diff
@@ -123,12 +124,26 @@ def applied_coupon
...
+
+      true
+    end
```
Comment (vincent-pochet) [inline app/services/invoices/create_service.rb:145]: You could make it a bit shorter:

```ruby
    def should_create_applied_prepaid_credit?
       return false unless wallet&.active?
       
       wallet.balance > 0
    end
```

Comment (lovrocolic) [reply app/services/invoices/create_service.rb]: refactored

File: app/services/credits/applied_prepaid_credit_service.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+      super(nil)
+    end
+
+    def create
+      return result if already_applied?
+
+      amount_cents = compute_amount
+      amount = compute_amount_from_cents(amount_cents)
+      credit_amount = amount.fdiv(BigDecimal(wallet.rate_amount))
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/credits/applied_prepaid_credit_service.rb]: fixed

File: app/services/wallets/balance/decrease_service.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        wallet.update!(
+          balance: wallet.balance - amount,
+          credits_balance: wallet.credits_balance - credits_amount,
+          last_balance_sync_at: Time.zone.now,
+          consumed_credits: wallet.consumed_credits + credits_amount,
+          consumed_amount: wallet.consumed_amount + amount,
+          last_consumed_credit_at: Time.current
+        )
+      end
```
Comment (vincent-pochet) [inline app/services/wallets/balance/decrease_service.rb:24]: Maybe we should return a result here. What do you think?

```ruby
  result.wallet = wallet
  result
rescue ActiveRecord::RecordInvalid => e
  result.fail_with_validations!(e.record)
```

File: app/services/credits/applied_prepaid_credit_service.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    end
+
+    def compute_amount_from_cents(amount)
+      currency = invoice.amount.currency
+
+      amount.round.fdiv(currency.subunit_to_unit)
+    end
+
+    def balance_cents
+      balance = BigDecimal(wallet.balance)
```
Comment (vincent-pochet) [inline app/services/credits/applied_prepaid_credit_service.rb:63]: Do we still need this one?

Comment (lovrocolic) [reply app/services/credits/applied_prepaid_credit_service.rb]: leftover:) Removed

File: app/services/wallets/balance/decrease_service.rb (hunk 4 above)
Comment (lovrocolic) [reply app/services/wallets/balance/decrease_service.rb:27]: wallet is actually not used in a caller for now, but added it if that's the standard :)

**Pull request merged**